upup/pkg/fi/cloudup/openstacktasks/sshkey.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstacktasks
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/keypairs"
+
+	"k8s.io/kops/pkg/pki"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/openstack"
+)
+
+//go:generate fitask -type=SSHKey
+type SSHKey struct {
+	Name      *string
+	Lifecycle *fi.Lifecycle
+
+	PublicKey *fi.ResourceHolder
+
+	KeyFingerprint *string
+}
+
+var _ fi.CompareWithID = &SSHKey{}
+
+func (e *SSHKey) CompareWithID() *string {
+	return e.Name
+}
+
+func (e *SSHKey) Find(c *fi.Context) (*SSHKey, error) {
+	cloud := c.Cloud.(openstack.OpenstackCloud)
+	rs, err := cloud.ListKeypair(openstackKeyPairName(fi.StringValue(e.Name)))
+	if err != nil {
+		return nil, err
+	}
+	if rs == nil {
+		return nil, nil
+	}
+	actual := &SSHKey{
+		Name:           e.Name,
+		KeyFingerprint: fi.String(rs.Fingerprint),
+	}
+
+	// Avoid spurious changes
+	if fi.StringValue(actual.KeyFingerprint) == fi.StringValue(e.KeyFingerprint) {
+		glog.V(2).Infof("SSH key fingerprints match; assuming public keys match")
+		actual.PublicKey = e.PublicKey
+	} else {
+		glog.V(2).Infof("Computed SSH key fingerprint mismatch: %q %q", fi.StringValue(e.KeyFingerprint), fi.StringValue(actual.KeyFingerprint))
+	}
+	actual.Lifecycle = e.Lifecycle
+	return actual, nil
+}
+
+func (e *SSHKey) Run(c *fi.Context) error {
+	if e.KeyFingerprint == nil && e.PublicKey != nil {
```

Comment:
Sorry, not quite understand. why need compute 'KeyFingerprint' ? The render function below will compute it.

Reply:
this is almost copy paste from awstasks/sshkey.go, there it is in both places as well. Maybe there is some reason, I have no idea

Reply:
Probably because AWS has a special way to calculate fingerprints. (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html#verify-key-pair-fingerprints)

Reply:
ok, then it is not needed. I deleted this part and it works well after that.